napi_js_data.md

```diff
@@ -0,0 +1,760 @@
+
+Fundamentally, these APIs are used to do one of the following:
+1. Create a new JavaScript object
+2. Convert from a primitive C++ type to N-API value
+3. Convert from N-API value to a primitive C++ type
+
+N-API values are represented by the type `napi_value`.
+Any N-API API that requires a JavaScript object takes in a `napi_value`.
+In some cases, the API does check the type of the `napi_value` up-front.
+However, for better performance, it's better that the caller makes sure that 
+the `napi_value` in question is of the JavaScript type expected by the API.
+
+## Enum types
+
+### *napi_value_type*
+
+#### Definition
+```
+enum napi_valuetype {
+  napi_undefined,
+  napi_null,
+  napi_boolean,
+  napi_number,
+  napi_string,
+  napi_symbol,
+  napi_object,
+  napi_function,
+  napi_external,
+}
+```
+
+#### Description
+Describes the type of a napi_value. This generally corresponds with the types described
+in [Section 6.1 of the ECMAScript Language Specification](https://tc39.github.io/ecma262/#sec-ecmascript-language-types).
+In addition to types in that section, `napi_value_type` also can represent Functions and Objects with external data.
+
+### *napi_typedarray_type*
+
+#### Definition
+```
+enum napi_typedarray_type {
+  napi_int8_array,
+  napi_uint8_array,
+  napi_uint8_clamped_array,
+  napi_int16_array,
+  napi_uint16_array,
+  napi_int32_array,
+  napi_uint32_array,
+  napi_float32_array,
+  napi_float6_array4
+}
+```
+
+#### Description
+This represents the underlying binary scalar datatype of the TypedArray.
+Elements of this enum correspond to [Section 22.2 of the ECMAScript Language Specification](https://tc39.github.io/ecma262/#sec-typedarray-objects).
+
+## Functions
+
+### *napi_create_array*
+
+#### Signature
+```
+napi_status napi_create_array(napi_env env, napi_value* result)
+```
+
+#### Parameters
+- `[in]  env`: The environment that the N-API API is invoked under
+- `[out] result`: A `napi_value` representing a JavaScript Array
+
+#### Return value
+- `napi_ok` if the API succeeded.
+
+#### Description
+This API returns a N-API value corresponding to a JavaScript Array type. 
+JavaScript arrays are described in [Section 22.1](https://tc39.github.io/ecma262/#sec-array-objects) of the ECMAScript Language Specification.
+
+### *napi_create_array_with_length*
+
+#### Signature
+```
+napi_status napi_create_array(napi_env env, size_t length, napi_value* result)
+```
+
+#### Parameters
+- `[in]  env`: The environment that the API is invoked under
+- `[in]  length`: The initial length of the Array
+- `[out] result`: A `napi_value` representing a JavaScript Array
+
+#### Return value
+- `napi_ok` if the API succeeded.
+
+#### Description
+This API returns a N-API value corresponding to a JavaScript Array type. 
+The Array's length property is set to the passed in length parameter. 
+However, the underlying buffer is not guaranteed to be pre-allocated- that behavior
```

Comment:
The discussion of a "pre-allocated buffer" is confusing to me because the user of the array does not ever have to do anything to allocate a buffer for an array. (Unlike an arraybuffer or typed-array.)